hvplot/plotting/scatter_matrix.py

```diff
@@ -38,19 +49,74 @@ def scatter_matrix(data, c=None, chart='scatter', diagonal='hist', alpha=0.5, **
     diagonal = HoloViewsConverter._kind_mapping[diagonal]
     chart = HoloViewsConverter._kind_mapping[chart]
 
+    #remove datashade kwds
+    if datashade == True or rasterize == True:
+        ds_kwds = {}
+        if 'aggregator' in kwds:
+            ds_kwds['aggregator'] = kwds.pop('aggregator')
+        
+    #remove dynspread kwds
+    if dynspread:
+        dn_kwds = {}
+        if 'max_px' in kwds:
+            dn_kwds['max_px'] = kwds.pop('max_px')
+        if 'threshold' in kwds:
+            dn_kwds['threshold'] = kwds.pop('threshold')
+        if 'how' in kwds:
+            dn_kwds['how'] = kwds.pop('how')
+        if 'mask' in kwds:
+            dn_kwds['mask'] = kwds.pop('mask')
+
+    if spread:
+        sp_kwds = {}
+        if 'px' in kwds:
+            sp_kwds['px'] = kwds.pop('px')
+        if 'shape' in kwds:
+            sp_kwds['shape'] = kwds.pop('shape')
+        if 'how' in kwds:
+            sp_kwds['how'] = kwds.pop('how')
+        if 'mask' in kwds:
+            sp_kwds['mask'] = kwds.pop('mask')
+        
+
     colors = _hv.plotting.util.process_cmap('Category10', categorical=True)
     chart_opts = dict(alpha=alpha, cmap=colors, tools=['box_select', 'lasso_select'],
                       nonselection_alpha=0.1, **kwds)
 
     grid = _hv.operation.gridmatrix(data, diagonal_type=diagonal, chart_type=chart)
+
     if c:
         chart_opts['color_index'] = c
         grid = grid.map(lambda x: x.clone(vdims=x.vdims+[c]), 'Scatter')
         groups = _hv.operation.gridmatrix(data.groupby(c).overlay(),
                                           chart_type=chart,
                                           diagonal_type=diagonal)
-        grid = (grid * groups).map(lambda x: x.get(0) if isinstance(x.get(0), chart) else x.get(1),
-                                   _hv.Overlay)
+        grid = grid * groups
 
+    if datashade or rasterize:
+        if datashade:
+            grid = grid.map(lambda x: x.get(0)
+            .apply(_datashade, **ds_kwds) if isinstance(x.get(0), chart) else x.get(1),
+                                _hv.Overlay)
+        else:
+            grid = grid.map(lambda x: x.get(0)
+            .apply(_rasterize, **ds_kwds) if isinstance(x.get(0), chart) else x.get(1),
+                                _hv.Overlay)
+        if dynspread:
+            grid = grid.map(lambda x: x.get(0)
+            .apply(_dynspread, **dn_kwds) if isinstance(x.get(0), chart) else x.get(1),
+                                _hv.Overlay)
+        elif spread:
+            grid = grid.map(lambda x: x.get(0)
+            .apply(_spread, **sp_kwds) if isinstance(x.get(0), chart) else x.get(1),
+                                _hv.Overlay)
+    else:
+        grid = grid.map(lambda x: x.get(0) if isinstance(x.get(0), chart) else x.get(1),
+                                _hv.Overlay)
+    
```

Comment:
I haven't tested this code, but I'd find it easier to follow if it was pulled out as a named function (preferably with a docstring), so that I could follow the logic and parameters more easily. E.g.:

```suggestion
   def apply_to_charts(grid, op, kwds):
        return grid.map(lambda x: x.get(0).apply(op, **kwds)
                        if isinstance(x.get(0), chart)
                        else x.get(1), _hv.Overlay)
    
   if datashade or rasterize:
        if datashade: grid = apply_to_charts(_datashade, **ds_kwds)
        else:         grid = apply_to_charts(_rasterize, **ds_kwds)
        if dynspread: grid = apply_to_charts(_dynspread, **dn_kwds)
        elif spread:  grid = apply_to_charts(_spread,    **sp_kwds)
    else:
        grid = apply_to_charts(lambda x: x)
```